core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/SailSourceConnection.java

```diff
@@ -255,6 +253,35 @@ protected EvaluationStrategy getEvaluationStrategy(Dataset dataset, TripleSource
 		}
 	}
 
+	@Override
+	public TupleExpr explain(Query.QueryExplainLevel queryExplainLevel, TupleExpr tupleExpr, Dataset activeDataset,
+			BindingSet bindings, boolean includeInferred) {
+		switch (queryExplainLevel) {
+		case Executed:
+			try (CloseableIteration<? extends BindingSet, QueryEvaluationException> evaluate = evaluate(tupleExpr,
+					activeDataset, bindings, includeInferred)) {
+				while (evaluate.hasNext()) {
+					evaluate.next();
+				}
+			}
+
+			return tupleExpr;
+		case Optimized:
+			SailSource branch = branch(IncludeInferred.fromBoolean(includeInferred));
+			SailDataset rdfDataset = branch.dataset(getIsolationLevel());
+
+			TripleSource tripleSource = new SailDatasetTripleSource(vf, rdfDataset);
+			EvaluationStrategy strategy = getEvaluationStrategy(activeDataset, tripleSource);
+
+			return strategy.optimize(tupleExpr, store.getEvaluationStatistics(), bindings);
+
+		case Unoptimized:
+			return tupleExpr;
+		}
+
+		throw new UnsupportedOperationException("Unsupported queryExplainLevel: " + queryExplainLevel);
+	}
+
```

Comment:
Implementation of getting the query plan.

A few things to note:
 - we use the evaluate() method to run the query instead of duplicating the code. We might need to duplicate it though since we need access to the evaluation strategy to enable tracking of result size when that gets merged in.
 - when using evaluate() it would clone tupleExpr, that had to be removed if we don't want to duplicate a lot of code
 - the code in `Optimized` is copied from evaluate, not sure if this is want